File: backend/internal/shared/kernel/values/recurring_test.go
// SPDX-License-Identifier: BUSL-1.1
// SPDX-FileCopyrightText: 2026 Gradion

package values

// Both sides of the monthly reading answer the SAME corpus.
//
// The file this reads is also read by the browser bundle's own test, so a case
// added here is answered twice or one side goes red. Writing the cases out
// twice instead would prove only that the two lists agreed on the day somebody
// typed them, which is the arrangement the minor-unit vocabulary gate exists
// because of.

import (
	"encoding/json"
	"os"
	"testing"
)

type monthlyCase struct {
	Name        string `json:"name"`
	Currency    string `json:"currency"`
	AnnualMinor int64  `json:"annual_minor"`
	// MonthlyMinor and Approximate are what BOTH implementations must answer.
	MonthlyMinor int64 `json:"monthly_minor"`
	Approximate  bool  `json:"approximate"`
}

func TestMonthlyEquivalentAnswersTheSharedCorpus(t *testing.T) {
	t.Parallel()

	raw, err := os.ReadFile("testdata/monthly_equivalent.json")
	if err != nil {
		t.Fatalf("reading the shared corpus: %v", err)
	}
	var corpus struct {
		Cases []monthlyCase `json:"cases"`
	}
	if err := json.Unmarshal(raw, &corpus); err != nil {
		t.Fatalf("parsing the shared corpus: %v", err)
	}
	// A corpus that parsed to nothing agrees with every implementation,
	// including a wrong one. The floor is what stops a green result meaning
	// that the file moved.
	if len(corpus.Cases) < 8 {
		t.Fatalf("the shared corpus holds %d cases — too few to have been read correctly, and a "+
			"test that checks nothing passes over any arithmetic at all", len(corpus.Cases))
	}

	for _, c := range corpus.Cases {
		t.Run(c.Name, func(t *testing.T) {
			monthly, approximate := MonthlyEquivalent(c.AnnualMinor)
			if monthly != c.MonthlyMinor {
				t.Errorf("MonthlyEquivalent(%d) = %d, want %d (%s)",
					c.AnnualMinor, monthly, c.MonthlyMinor, c.Currency)
			}
			if approximate != c.Approximate {
				t.Errorf("MonthlyEquivalent(%d) approximate = %v, want %v",
					c.AnnualMinor, approximate, c.Approximate)
			}
		})
	}
}

// Twelve months add back to the year exactly when the division said they would.
// This is the property the approximate flag claims, checked against the flag
// rather than restated as more cases.
func TestAnExactMonthlyReadingMultipliesBackToTheYear(t *testing.T) {
	t.Parallel()
	for annual := int64(0); annual < 500; annual++ {
		monthly, approximate := MonthlyEquivalent(annual)
		exact := monthly*12 == annual
		if exact == approximate {
			t.Fatalf("MonthlyEquivalent(%d) = %d reports approximate=%v, but twelve of it %s the year",
				annual, monthly, approximate, map[bool]string{true: "reproduce", false: "do not reproduce"}[exact])
		}
	}
}

File: frontend/src/format/recurring.test.ts
import { readFileSync } from "node:fs";
import { describe, expect, it } from "vitest";

import { monthlyEquivalent } from "./recurring";

/**
 * The browser's monthly reading answers the SAME corpus the Go kernel does.
 *
 * The file is read from the backend tree on purpose. A copy of the cases here
 * would prove the two lists agreed the day somebody typed them and nothing
 * about the next edit, which is the drift the minor-unit vocabulary gate was
 * written after.
 */
type MonthlyCase = Readonly<{
  name: string;
  currency: string;
  annual_minor: number;
  monthly_minor: number;
  approximate: boolean;
}>;

// Relative to THIS file, so the reference survives the frontend being run from
// anywhere. It points into the backend tree deliberately: one corpus, two
// readers.
const CORPUS_PATH =
  "../../../backend/internal/shared/kernel/values/testdata/monthly_equivalent.json";

const cases: MonthlyCase[] = JSON.parse(
  readFileSync(new URL(CORPUS_PATH, import.meta.url), "utf8"),
).cases;

describe("monthlyEquivalent", () => {
  // A corpus that read as empty agrees with every implementation, including a
  // wrong one, so the floor runs before the cases do.
  it("reads the shared corpus the Go side reads", () => {
    expect(cases.length).toBeGreaterThanOrEqual(8);
  });

  it.each(cases)("$name", (c) => {
    const reading = monthlyEquivalent(c.annual_minor);
    expect(reading).not.toBeNull();
    expect(reading?.monthlyMinor).toBe(c.monthly_minor);
    expect(reading?.approximate).toBe(c.approximate);
  });

  // The property the approximate flag claims, checked rather than restated as
  // more corpus rows.
  it("reports exact only when twelve months add back to the year", () => {
    for (let annual = 0; annual < 500; annual += 1) {
      const reading = monthlyEquivalent(annual);
      expect(reading).not.toBeNull();
      expect((reading?.monthlyMinor ?? 0) * 12 === annual).toBe(
        !reading?.approximate,
      );
    }
  });

  // A figure this cannot divide exactly answers with nothing rather than a
  // guess. It returns null rather than throwing because the callers render it
  // beside an annual figure that displays correctly on its own, and a throw
  // would take the whole record page down over the derived line.
  it("answers nothing for a figure it cannot divide exactly", () => {
    expect(monthlyEquivalent(10.5)).toBeNull();
    expect(monthlyEquivalent(Number.NaN)).toBeNull();
    // Past the safe-integer bound the value arrived having already lost
    // precision in JSON, so there is no exact figure left to divide.
    expect(monthlyEquivalent(Number.MAX_SAFE_INTEGER + 2)).toBeNull();
  });
});
